mne/viz/_figure.py

```diff
@@ -2031,483 +308,197 @@ def _update_data(self):
-            this_z = self.mne.zorder['bads' if is_bad_ch else 'data']
-            if self.mne.butterfly and not is_bad_ch:
-                this_z = self.mne.zorder.get(this_type, this_z)
-            # plot each trace multiple times to get the desired epoch coloring.
-            # use masked arrays to plot discontinuous epochs that have the same
-            # color in a single plot() call.
-            if self.mne.instance_type == 'epochs':
-                this_colors = custom_colors[ii]
-                for cix, color in enumerate(np.unique(this_colors, axis=0)):
-                    bool_ixs = (this_colors == color).all(axis=1)
-                    mask = np.zeros_like(this_times, dtype=bool)
-                    _starts = self.mne.boundary_times[epoch_ix][bool_ixs]
-                    _stops = self.mne.boundary_times[epoch_ix + 1][bool_ixs]
-                    for _start, _stop in zip(_starts, _stops):
-                        _mask = np.logical_and(_start < this_times,
-                                               this_times <= _stop)
-                        mask = mask | _mask
-                    _times = np.ma.masked_array(this_times, mask=~mask)
-                    # always use the existing traces first
-                    if cix == 0:
-                        line.set_xdata(_times)
-                        line.set_ydata(this_data)
-                        line.set_color(color)
-                        line.set_zorder(this_z)
-                    else:  # make new traces as needed
-                        _trace = self.mne.ax_main.plot(
-                            _times, this_data, color=color, zorder=this_z,
-                            **self.mne.trace_kwargs)
-                        self.mne.epoch_traces.extend(_trace)
-            else:
-                line.set_xdata(this_times)
-                line.set_ydata(this_data)
-                line.set_color(ch_colors[ii])
-                line.set_zorder(this_z)
-        # update xlim
-        self.mne.ax_main.set_xlim(*time_range)
-        # draw scalebars maybe
-        if self.mne.scalebars_visible:
-            self._show_scalebars()
-        # redraw event lines
-        if self.mne.event_times is not None:
-            self._draw_event_lines()
-
-    def _redraw(self, update_data=True, annotations=False):
-        """Redraw (convenience method for frequently grouped actions)."""
-        if update_data:
-            self._update_data()
-        if self.mne.vline_visible and self.mne.is_epochs:
-            # prevent flickering
-            _ = self._recompute_epochs_vlines(None)
-        self._draw_traces()
-        if annotations and not self.mne.is_epochs:
-            self._draw_annotations()
-        self.canvas.draw_idle()
+    def _get_epoch_num_from_time(self, time):
+        epoch_nums = self.mne.inst.selection
+        return epoch_nums[np.searchsorted(self.mne.boundary_times[1:], time)]
+
+    def _redraw(self, **kwargs):
```

Comment:
Missing `@abstractmethod`?

Reply:
At first I thought this is not necessary for pyqtgraph but now I got it too so it can be abstract.